platform/commonUI/edit/src/representers/EditToolbar.js

```diff
@@ -156,23 +144,17 @@ define(
         };
 
         /**
-         * Set the current selection. Visibility of sections
-         * and items in the toolbar will be updated to match this.
-         * @param {Array} s the new selection
+         * Updates the toolbar with a new structure.
+         *
+         * @param {Array} structure the toolbar structure
+         * @param {Array} selection the new selected control in the toolbar
          */
-        EditToolbar.prototype.setSelection = function (s) {
+        EditToolbar.prototype.updateToolbar = function (structure, selection) {
             var self = this;
 
-            // Show/hide controls in this section per applicability
-            function refreshSectionApplicability(section) {
-                var count = 0;
-                // Show/hide each item
-                (section.items || []).forEach(function (item) {
-                    item.hidden = !self.isApplicable(item);
-                    count += item.hidden ? 0 : 1;
-                });
-                // Hide this section if there are no applicable items
-                section.hidden = !count;
+            // Show/hide controls per applicability
+            function refreshApplicability(item) {
```

Comment:
Applicability should be handled in the toolbar provider's `toolbar` method: it should only return controls that are valid for a given selection.